### PR TITLE
Logic: Improve `FindCommand`

### PR DESCRIPTION
**Pull request by lycjackie**
The `FindCommand` only allow searching of `Name` keywords. However there are 
other field which the user may be interested in searching. Such as
- `Code`
- `Credits`

To allow our user to find module more effectively in `PWE`

Let's improve on our `FindCommand` and:
- allow finding of `Name` keywords with `PREFIX_NAME`
- allow finding of `Code` keywords with `PREFIX_CODE`
- allow finding of `Credits` keywords with `PREFIX_CREDITS`
- refractor existing unit tests and append with `PREFIX_NAME`
- add new unit tests for finding of `Code` keywords
- add new unit tests for finding of `Credits` keywords

Commits for easy reference
* [ 1/23] [Model/module: add new abstract class `KeywordsPredicate`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/129e5ba3b8254d2275590a481e833b0f747f08d6)
* [ 2/23] [Model: make `NameContainsKeywordsPredicate` extends `KeywordsPredicate](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/042762de7b364ac8ba641fa2ad5c352a687c37d0)
* [ 3/23] [Logic: make `FindCommandParser#parse(...)` parse `PREFIX_NAME`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/3d2810020e9d2ed529a3b6e3e9c376f0a36954e3)
* [ 4/23] [commands/FindCommandTest: refactor methods name](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/2969187abf0cb2ab8dabcfcc1d12df921c5d4119)
* [ 5/23] [systemtests/FindCommandSystemTest: update broken unit tests in `find()`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/1c62d34916f94bd732e1905e37ffb135b31e3626)
* [ 6/23] [test/parser: update `AddressBookParserTest` broken unit test](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/6e948b98d7ea990f11cbd76953ca7a4cf4d490c9)
* [ 7/23] [test/parser: update `FindCommandParserTest` broken unit test](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/51d2926e90fd28b4504a54b6c10f19022d4c57a0)
* [ 8/23] [test/parser: refactor ` FindCommandParserTest` attribute name](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/d90a0b14667ae23f4784c32f1c2c2396720d9248)
* [ 9/23] [Model/module: add `CodeContainsKeywordsPredicate` class](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/8586867f0c798fffbade680f4a0b00f5d89b3521)
* [10/23] [Logic: make `FindCommandParser#parse(...)` parse `PREFIX_CODE`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/9d1cdd95d8bdb238cd319fa7adbe4e939564314c)
* [11/23] [FindCommand: update `MESSAGE_USAGE`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/d2cf37e5c797781de48fbc22c921b92d2fac55fd)
* [12/23] [test/module: add unit tests to `CodeContainsKeywordsPredicateTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/71fe8a74d07cecaf877e362b966251ed0d078b5e)
* [13/23] [test/parser: add unit test to `FindCommandParserTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/f90c2f6507c4dba6e453d73ede9627e0c6af9fdb)
* [14/23] [test/commands: add unit test to `FindCommandTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/554eaf9cf01ebbc048f0a24bae26f682b9818643)
* [15/23] [systemtests/FindCommandSystemTest: add unit tests](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/c8ffe342107e79a1a8824837d235983c97f59a16)
* [16/23] [Model/module: add `CreditsContainsKeywordsPredicate` class](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/44bbfcdb1b35fa4225e3ea4a893288fec9d5f2a8)
* [17/23] [Logic: make `FindCommandParser#parse(...)` parse `PREFIX_CREDITS`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/623e551f2e1663713f261d7a275d47d406adab19)
* [18/23] [FindCommand: update `MESSAGE_USAGE`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/3b52d53ae742bd434337d8fec5d0a2828bc239e7)
* [19/23] [Logic: refactor `FindCommandParser#parse(...)`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/c57229b3f625bf4da8b1b4ebe7386d2dc9e8ad80)
* [20/23] [test/module: add unit tests to `CreditsContainsKeywordsPredicateTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/04dd0986fbe1bb02dcda4cd097eb6e31442036a4)
* [21/23] [test/parser: add unit test to `FindCommandParserTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/ea05da395f45d6f5a43041dbbfb869608bcde976)
* [22/23] [test/commands: add unit test to `FindCommandTest`](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/ae7d847a10524168d23febcacf50f25ff1519ed5)
* [23/23] [systemtests/FindCommandSystemTest: add unit test](https://github.com/cs2113-ay1819s2-t09-1/main/pull/35/commits/171e28c69b84fc20debc442a8b4d39c35d0c5d8a)

### CODE REVIEW

File: src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        // same values -> returns true
+        CodeContainsKeywordsPredicate firstPredicateCopy =
+                new CodeContainsKeywordsPredicate(firstPredicateKeywordList);
+        assertTrue(firstPredicate.equals(firstPredicateCopy));
+
+        // different types -> returns false
+        assertFalse(firstPredicate.equals(1));
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
```
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java]: As mention in #20, 

The purpose of this PR is to improve `FindCommand`, we will try to improve on this issue on another PR

File: src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        // same values -> returns true
+        CreditsContainsKeywordsPredicate firstPredicateCopy =
+                new CreditsContainsKeywordsPredicate(firstPredicateKeywordList);
+        assertTrue(firstPredicate.equals(firstPredicateCopy));
+
+        // different types -> returns false
+        assertFalse(firstPredicate.equals(1));
+
+        // null -> returns false
+        assertFalse(firstPredicate.equals(null));
```
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java]: As mention in #20, 

The purpose of this PR is to improve `FindCommand`, we will try to improve on this issue on another PR

File: src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java]: As mention in #20 

We might need more investigation with regards to these issues.

I will send in another PR to close #20, as such I will be resolving this.

File: src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java]: As mention in #20 

We might need more investigation with regards to these issues.

I will send in another PR to close #20, as such I will be resolving this.

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all modules whose names contain any of "
-            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
-            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
-            + "Example: " + COMMAND_WORD + " alice bob charlie";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all modules whose names or code "
+            + "contain any of the specified keywords (case-insensitive)"
+            + " and displays them as a list with index numbers.\n"
+            + "Parameters: "
+            + "[" + PREFIX_NAME + "NAME name...]"
+            + "[" + PREFIX_CODE + "Code code...]\n"
```
Comment (Creastery) [inline src/main/java/seedu/address/logic/commands/FindCommand.java]: I think this should be:
```java
+ "[" + PREFIX_CODE + "CODE code...]\n"
```

Comment (Creastery) [inline src/main/java/seedu/address/logic/commands/FindCommand.java]: I think this should be:
```java
+ "[" + PREFIX_CODE + "CODE code...]\n"
```

**Review by Creastery — CHANGES_REQUESTED** [review]
Overall, really good improvements made to `FindCommand`!

Just a couple of things to amend:

```
[ 3/23] Logic: make FindCommandParser#parse(...) parse PREFIX_NAME
```
Since this commit changes the `find` command usage (i.e. `find n/abc` instead of `find abc`) for parsing the arguments, you also broke the existing unit tests at that point.

Please update the commit message to include a disclaimer that existing unit tests will break with the change in command usage, and will be fixed later on.

---

```
[19/23] Logic: refactor FindCommand#parse
```
Commit title should have been:

```
[19/23] Logic: refactor `FindCommandParser#parse(...)`
```

File: src/main/java/seedu/address/logic/commands/FindCommand.java (hunk 3 above)
Comment (lycjackie) [reply src/main/java/seedu/address/logic/commands/FindCommand.java]: Thanks for noticing this!

Fixed

**Comment by lycjackie** [comment]
> Overall, really good improvements made to `FindCommand`!
> 
> Just a couple of things to amend:
> 
> ```
> [ 3/23] Logic: make FindCommandParser#parse(...) parse PREFIX_NAME
> ```
> Since this commit changes the `find` command usage (i.e. `find n/abc` instead of `find abc`) for parsing the arguments, you also broke the existing unit tests at that point.
> 
> Please update the commit message to include a disclaimer that existing unit tests will break with the change in command usage, and will be fixed later on.
> 
> ```
> [19/23] Logic: refactor FindCommand#parse
> ```
> Commit title should have been:
> 
> ```
> [19/23] Logic: refactor `FindCommandParser#parse(...)`
> ```

Fixed all mention in the latest push.

File: src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CreditsContainsKeywordsPredicateTest.java]: As mention in #20

We might need more investigation with regards to these issues.

As suggested previously, we will leave code quality improvement to another PR

File: src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lycjackie) [reply src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java]: As mention in #20

We might need more investigation with regards to these issues.

As suggested previously, we will leave code quality improvement to another PR

[the remark this answers is not in the document]
Comment (Creastery) [reply src/test/java/seedu/address/model/module/CodeContainsKeywordsPredicateTest.java:37]: Closing in view of #20.

**Review by Creastery — APPROVED** [review]

**Comment by Creastery** [comment]
LGTM. Outstanding Codacy issues will be resolved at a later date as per #20. Approving and merging. 
Thanks for the quality work!

**Pull request merged**